internal/api/hooks.go

```diff
@@ -55,20 +74,219 @@ func (a *API) runHook(ctx context.Context, tx *storage.Connection, name string,
 	return response, nil
 }
 
-// invokeHook invokes the hook code. tx can be nil, in which case a new
+func (a *API) runHTTPHook(hookConfig conf.ExtensibilityPointConfiguration, input, output any) ([]byte, error) {
+
+	client := http.Client{
+		Timeout: DefaultHTTPHookTimeout,
+	}
+	requestURL := hookConfig.URI
+	hookLog := logrus.WithFields(logrus.Fields{
+		"component": "auth_hook",
+		"url":       requestURL,
+	})
+
+	inputPayload, err := json.Marshal(input)
+	if err != nil {
+		return nil, err
+	}
+
+	if isOverSizeLimit(inputPayload) {
+		return nil, internalServerError("Over size limit")
+	}
+	startTime := time.Now()
+
+	for i := 0; i < DefaultHTTPHookRetries; i++ {
+		hookLog.Infof("invocation attempt: %d", i)
+		msgID := uuid.Must(uuid.NewV4())
+		currentTime := time.Now()
+		signatureList, err := generateSignatures(hookConfig.HTTPHookSecrets, msgID, currentTime, inputPayload)
+		if err != nil {
+			return nil, err
+		}
+
+		req, err := http.NewRequest(http.MethodPost, requestURL, bytes.NewBuffer(inputPayload))
+		if err != nil {
+			return nil, internalServerError("Failed to make request object").WithInternalError(err)
+		}
+
+		req.Header.Set("Content-Type", "application/json")
+
+		req.Header.Set("webhook-id", msgID.String())
+		req.Header.Set("webhook-timestamp", fmt.Sprintf("%d", currentTime.Unix()))
+		req.Header.Set("webhook-signature", strings.Join(signatureList, ", "))
+
+		watcher, req := watchForConnection(req)
+		rsp, err := client.Do(req)
+
+		if err != nil {
+
+			if terr, ok := err.(net.Error); ok && terr.Timeout() {
+				hookLog.Errorf("Request timed out for attempt %d with err %s", i, err)
+				// TODO: workshop the sleep time value
+				time.Sleep(DefaultHTTPHookTimeout)
+				continue
+			} else if !watcher.gotConn && i < DefaultHTTPHookRetries-1 {
+				hookLog.Errorf("Failed to establish a connection on attempt %d with err %s", i, err)
+				time.Sleep(DefaultHTTPHookTimeout)
+				continue
+			} else if i == DefaultHTTPHookRetries-1 {
+				return nil, httpError(http.StatusGatewayTimeout, "Failed to perform webhook in time frame (%v seconds)", time.Since(startTime).Seconds())
```

Comment:
Check if gatway timeout is the correct HTTP code.

Reply:
Double checked with [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504) and does seem to tally with what we wish to indicate-  we can't reach the server after two tries. This is also what was [used in the previous implementation](https://github.com/supabase/gotrue/blob/j0/update_otel_deps/internal/api/hooks.go#L105)